bcda/cclf/importer.go

```diff
@@ -1,83 +1,113 @@
 package cclf
 
 import (
+	"bufio"
+	"bytes"
 	"context"
-	"database/sql"
 	"fmt"
 
 	"github.com/CMSgov/bcda-app/bcda/cclf/metrics"
-	"github.com/CMSgov/bcda-app/bcda/models"
-	"github.com/lib/pq"
-	"github.com/pkg/errors"
-	"github.com/sirupsen/logrus"
+	"github.com/jackc/pgx"
+	"github.com/jackc/pgx/pgtype"
 )
 
 // A cclf8Importer is not safe for concurrent use by multiple goroutines.
 // It should be scoped to a single *sql.Tx
 type cclf8Importer struct {
-	logger *logrus.Logger
+	ctx context.Context
 
-	inprogress *sql.Stmt
+	scanner        *bufio.Scanner
+	reportInterval int
+	cclfFileID     uint // CCLFFile ID that will be associated with all created benes
 
-	pendingQueries    int
-	maxPendingQueries int
+	importCount   int
+	processedMBIs map[string]struct{}
 }
 
-func (cclfImporter *cclf8Importer) do(ctx context.Context, tx *sql.Tx, bene models.CCLFBeneficiary) error {
-	if cclfImporter.inprogress == nil {
-		if err := cclfImporter.refreshStatement(ctx, tx); err != nil {
-			return errors.Wrap(err, "failed to refresh statement")
+func (importer *cclf8Importer) Next() bool {
+	// Loops through the scanner until we either:
+	// 1. Encounter the end of the data
+	// 2. Find an MBI that we have not processed yet
+	//
+	// This logic exists in the Next() function because it
+	// is the only way for us to ignore an already ingested MBI.
+	// If we made this check in Values() and return an error or
+	// return empty data, then the copy will fail.
+	// NOTE: This choice was based on pgx v3.1.0.
+	//
+	// Since we're using the bufio.Scanner, we can make multiple calls
+	// to scanner.Bytes() without advancing the cursor.
+	for {
+		hasNext := importer.scanner.Scan()
+		if !hasNext {
+			return hasNext
 		}
-	}
 
-	if cclfImporter.pendingQueries >= cclfImporter.maxPendingQueries {
-		if err := cclfImporter.flush(ctx); err != nil {
-			return errors.Wrap(err, "failed to flush statement")
-		}
-		if err := cclfImporter.refreshStatement(ctx, tx); err != nil {
-			return errors.Wrap(err, "failed to refresh statement")
+		mbi := importer.getMBI()
+		// We've already processed this MBI before
+		if _, found := importer.processedMBIs[mbi]; found {
+			continue
```

Comment:
If you think we should follow this pattern for existence of a value in map, we can change the rest of the code base to match.

Reply:
I think it's a good pattern to use, but not so passionate that it needs to be enforced in all use cases. I find it clearer but understand that the map[interface{}]bool also makes sense for existence.